src/Main.elm

```diff
@@ -1,10 +1,571 @@
+                        Just parentId ->
+                            { parent = getParentFromDict parentId, child = Dir id audienceFolder }
+
+                        Nothing ->
+                            { parent = Nothing, child = Dir id audienceFolder }
+
+                File id audience ->
+                    case audience.folder of
+                        Just parentId ->
+                            { parent = getParentFromDict parentId, child = File id audience }
+
+                        Nothing ->
+                            { parent = Nothing, child = File id audience }
+
+                Root ->
+                    -- This state would never be reached. Given just to make compiler happy
+                    { parent = Nothing, child = File 0 defaultAudience }
+
+        getParentFromDict : Id -> Maybe Node
+        getParentFromDict parentId =
+            Dict.get parentId allNodesInDict
+
+        listOfParentChild : List { parent : Maybe Node, child : Node }
+        listOfParentChild =
+            allNodes
+                |> List.map (\node -> getParentChildNodes node)
+
+        -- This is flat tree with no parent-child relation yet
+        flatTree : Tree Node
+        flatTree =
+            allNodes
+                |> List.map Tree.singleton
+                |> Tree.tree Root
+    in
+    flatTreeToParentChildTree listOfParentChild (Just flatTree)
+
+
+
+-- This will order nodes in parent-child hierarchy
+
+
+flatTreeToParentChildTree : List { parent : Maybe Node, child : Node } -> Maybe (Tree Node) -> Tree Node
+flatTreeToParentChildTree listOfParentChild tree =
+    case listOfParentChild of
+        parentChild :: rest ->
+            case parentChild.parent of
+                Just parent ->
+                    flatTreeToParentChildTree rest (reorderNode { parent = parent, child = parentChild.child } tree)
+
+                Nothing ->
+                    flatTreeToParentChildTree rest tree
+
+        [] ->
+            Maybe.withDefault defaultTree tree
+
+
+reorderNode : { parent : Node, child : Node } -> Maybe (Tree Node) -> Maybe (Tree Node)
+reorderNode parentChild tree =
+    let
+        -- subTree is the child node
+        subTree : Maybe (Tree Node)
+        subTree =
+            findASubTree tree parentChild.child
+    in
+    -- tree is flat with no parent child hierarchy yet
+    tree
+        -- remove the child node from flat tree because we will insert it to parent soon
+        |> removeASubTree parentChild.child
+        -- and insert the subTree(child) to its parent tree
+        |> appendASubTreeAsChild subTree parentChild
+
+
+appendASubTreeAsChild : Maybe (Tree Node) -> { parent : Node, child : Node } -> Maybe (Tree Node) -> Maybe (Tree Node)
+appendASubTreeAsChild subTree parentChild tree =
+    let
+        appendTreeToTree tree_ =
+            Tree.appendChild (Maybe.withDefault defaultTree subTree) tree_
+    in
+    Zipper.fromTree (Maybe.withDefault defaultTree tree)
```

Comment:
You're doing `Zipper.fromTree` and `Zipper.toTree` in many of these functions. It could be nicer to make your functions work on `Zipper` and only convert trees from/to zippers at the very boundaries of these pipelines.